File: Iter2/src/LabCourse.java
import java.util.ArrayList;

public class LabCourse extends Course {

  private int sectionNumber;
	private int capacity;
	private DepartmentCourse mainCourse;
	private TeachingAssistant lecturer;

	public LabCourse(DepartmentCourse mainCourse) {
		this.mainCourse = mainCourse;
		this.lecturer = mainCourse.getTeachingAssistant(mainCourse.labSectionsCount());

		if (this.lecturer != null) {
			this.lecturer.assignLabCourse(this);
		}

		this.sectionNumber = mainCourse.labSectionsCount() + 1;
		this.courseId = mainCourse.getCourseId() + "." + this.sectionNumber;
		this.courseSchedule = new ArrayList<>();
	}

	public void addStudent(Student student) {
		if (this.capacity > 0) {
			registeredStudents.add(student);
			this.capacity--;
		}
	}

	public void setCapacity(int capacity) {
		this.capacity = capacity;
	}

	public int getCapacity() {
		return this.capacity;
	}

	public boolean hasCapacity() {
		return this.capacity > 0;
	}

	public void decreaseCapacity() {
		this.capacity--;
	}

	public int getSectionNumber() {
		return this.sectionNumber;
	}

	public DepartmentCourse getMainCourse() {
		return this.mainCourse;
	}

	public TeachingAssistant getLecturer() {
		return this.lecturer;
	}

	public ArrayList<Student> getRegisteredStudents() {
		return this.registeredStudents;
	}

}

File: Iter3/src/lab_course.py
import course
from department_course import DepartmentCourse
from teaching_assistant import TeachingAssistant
import student


class LabCourse(course.Course):
    __mainCourse: DepartmentCourse
    __lecturer: TeachingAssistant
    __sectionNumber: int
    __capacity: int

    def __init__(self, main_course: DepartmentCourse):
        self.__mainCourse = main_course

        self.__lecturer = main_course.get_teaching_assistant(main_course.lab_sections_count())
        if self.__lecturer is not None:
            self.__lecturer.assign_lab_course(self)

        self.__sectionNumber = main_course.lab_sections_count() + 1
        self.courseID = main_course.get_course_id() + "." + str(self.__sectionNumber)
        self.courseSchedule = []

    def add_student(self, student: student.Student):
        if self.__capacity > 0:
            self.__registeredStudents.append(student)
            self.__capacity = self.__capacity - 1

    def set_capacity(self, capacity: int):
        self.__capacity = capacity

    def has_capacity(self):
        return self.__capacity > 0

    def decrease_capacity(self):
        self.__capacity = self.__capacity - 1

    def get_main_course(self):
        return self.__mainCourse

    def get_lecturer(self):
        return self.__lecturer

    def get_registered_students(self):
        return self.__registeredStudents

    def get_section_number(self):
        return self.__sectionNumber

    def get_capacity(self):
        return self.__capacity
